### PR TITLE
1621/Listing Views

### PR DESCRIPTION
**Pull request by seanmalbert**
# Pull Request Template

## Issue
#1621 

Addresses # (issue)

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This adds a concept of "views" discussed within the issue.

Things to consider while reviewing:
- You can see that DetailView isn't used here. Specifying that many fields is tedious so I created FullView, which uses leftJoinAndSelect. I decided to keep DetailView in place, at least for this, so you can see how the configs extend each other.
- I also thought about how the configs extend each other. Instead of doing it by merging the fields, we could do it in `getView`, by calling super.getView(), which would add the extended views selects and joins to `qb`. Then we would add on other "view" configs. This may be a better approach
- Between the leftJoin and leftJoinAndSelect definitions, you can see they're handled differently and I think we should choose one for consistency. 
- I thought about having the views themselves create the query builder, instead of passing it in, but that made it much more difficult to test, and the services already have the repository injected, so it's easy to pass in.
- I didn't create a view factory here to generate the views, since we don't have many use cases at the moment, but that is something we could do, as @avaleske suggested in the issue

## How Can This Be Tested/Reviewed?

Both public and partners should work as before, but the list view should be considerably smaller.

### CODE REVIEW

**Comment by abbiefarr** [comment]
@seanmalbert thanks for putting this together!

I'm running into a problem in #1670, where the left joins I add overwhelm the service and cause the site functionality to seriously degrade. If I remove a few other left joins it works just fine, so I think we're just pushing the limit of the query.

I think the views drafted here would unblock the units summary progress. Is there an expected timeline for the remainder of this work?

**Comment by seanmalbert** [comment]
> @seanmalbert thanks for putting this together!
> 
> I'm running into a problem in #1670, where the left joins I add overwhelm the service and cause the site functionality to seriously degrade. If I remove a few other left joins it works just fine, so I think we're just pushing the limit of the query.
> 
> I think the views drafted here would unblock the units summary progress. Is there an expected timeline for the remainder of this work?

Hi @abbiefarr ,
Thanks for checking in on this, I haven't had time to get back to it. I think the remainder of this piece is to make sure the extension model works well by creating a view for getFullyJoinedQueryBuilder, which I can get to today. If your team can plan on reviewing this tomorrow to make sure it meets your needs as well, we should be able to get this in this week. Does that work for you?

**Comment by abbiefarr** [comment]
> If your team can plan on reviewing this tomorrow to make sure it meets your needs as well, we should be able to get this in this week. Does that work for you?

Sounds great! @avaleske is out of office this week, so feel free to add me and @willrlin as reviewers.

**Comment by seanmalbert** [comment]
@abbiefarr and @willrlin , this is ready for you to look at. I updated the PR description so please have a look there. For some reason I was only able to assign Abbie to this PR and Will didn't show up as an option 🤷

File: backend/core/src/listings/listings.service.ts [hunk 1]
```diff
@@ -77,19 +54,24 @@ export class ListingsService {
       // join on the listings we want to show.
       innerFilteredQuery.offset(offset).limit(params.limit as number)
     }
-
-    let listings = await this.getFullyJoinedQueryBuilder()
+    const view = new BaseView(this.listingRepository.createQueryBuilder("listings"))
+    // let listings = await view
```
Comment (emilyjablonski) [inline backend/core/src/listings/listings.service.ts:58]: Leftover comment

File: backend/core/src/listings/listings.service.ts [hunk 2]
```diff
@@ -173,7 +155,8 @@ export class ListingsService {
   }
 
   async findOne(listingId: string) {
-    const result = await this.getFullyJoinedQueryBuilder()
+    const view = new FullView(this.listingRepository.createQueryBuilder("listings")).getView()
+    const result = await view
       .where("listings.id = :id", { id: listingId })
       .orderBy({
         "preferences.ordinal": "ASC",
```
Comment (emilyjablonski) [inline backend/core/src/listings/listings.service.ts:162]: Ordering by preferences here feels weird?

File: backend/core/src/listings/views/config.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
+import { Views } from "./types"
+
+function getBaseAddressSelect(schemas: string[]): string[] {
+  const fields = ["city", "state", "street", "zipCode", "latitude", "longitude"]
```
Comment (emilyjablonski) [inline backend/core/src/listings/views/config.ts:4]: We also have `street2` just want to me sure this was a cognizant omission

File: backend/core/src/listings/views/index.spec.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    it("should call getView qb select and leftJoin", () => {
+      const view = new BaseView(mockListingsRepo.createQueryBuilder())
+
+      view.getView()
+
+      expect(mockQueryBuilder.select).toHaveBeenCalledTimes(1)
+      expect(mockQueryBuilder.leftJoin).toHaveBeenCalledTimes(6)
+    })
+
+    it("should cal getView qb leftJoinAndSelect", () => {
```
Comment (emilyjablonski) [inline backend/core/src/listings/views/index.spec.ts:65]: `call`

**Comment by abbiefarr** [comment]
Looks good from Detroit's perspective, thanks!

**Comment by willrlin** [comment]
I haven't been following the View discussion too deeply, so I trust @abbiefarr's review.

**Comment by seanmalbert** [comment]
@abbiefarr ,
Heads up, I realized I didn't include a way to pass the view as a query parameter, so I updated this so that ?view=<view> can get passed in for both the listing list and by id endpoints. I'm just making sure the pipeline passes and I'll merge in.

**Pull request merged**